Initialize a Rush workspace for client SDKs

bsiegel opened the pull request:
This PR creates a new Rush workspace, currently only containing the client SDKs. It includes several dependency version updates to align them across projects - this way, Rush can ensure that dependency versions for all of our packages remain in sync going forward.

Rush is configured to use pnpm as the package manager which provides several additional benefits. However Rush is not 100% compatible with the current version of pnpm (v3) so we should hold off merging this until that compatibility is added, or roll back to v2 of pnpm in rush.json

weshaggard commented:
cc @Azure/azure-sdk-eng

bterlson commented:
Some notes:

* `main` key of rollup-plugin-node-resolve is deprecated. Use mainFields instead.
* Seems like we have multiple versions of TypeScript. Probably not a major issue for now.
* `rush test` rebuilds the world after a `rush build`. Is this expected?
* `rush add` should really allow me to add local, unpublished packages (see: https://github.com/Microsoft/web-build-tools/issues/991)
* Instructions for taking an unpublished dependency seems important. Process is, basically, manually update your package.json with the dependency and run `rush update`.
* VSCode seems to stomp on some things, resulting in permissions errors when doing e.g. `rush update`. I suspect it's automatic type acquisition. I think I've gotten to a good place by enabling "typescript.disableAutomaticTypeAcquisition" in my workspace, might be worth adding that suggestion to the readme.

bterlson commented:
Played around with this (and built on it with #2373) and I like it lots. Happy to see this merged ASAP so we can start more refactorings like #2373.

bsiegel commented:
> * `main` key of rollup-plugin-node-resolve is deprecated. Use mainFields instead.
Fixed.

> * Seems like we have multiple versions of TypeScript. Probably not a major issue for now.
This was always true. API Extractor v6 requires an old version. If we upgrade to v7 it will resolve this issue.

> * `rush test` rebuilds the world after a `rush build`. Is this expected?
No, it's because we have many packages where the "test" script runs "build". We should clean this up in a followup PR.

> * `rush add` should really allow me to add local, unpublished packages (see: [Microsoft/web-build-tools#991](https://github.com/Microsoft/web-build-tools/issues/991))
😿 but for what it's worth you can edit the package.json manually and do `rush update` in this situation.

> * Instructions for taking an unpublished dependency seems important. Process is, basically, manually update your package.json with the dependency and run `rush update`.
Good catch! Fixed.

> * VSCode seems to stomp on some things, resulting in permissions errors when doing e.g. `rush update`. I suspect it's automatic type acquisition. I think I've gotten to a good place by enabling "typescript.disableAutomaticTypeAcquisition" in my workspace, might be worth adding that suggestion to the readme.
Added to the README. What are your feelings on checking in the `.vscode/settings.json` file with this change? I can see other settings being appropriate to commit as well, to help enforce some things repo-wide. If we want to make sure people don't keep getting diffs by making settings changes to their local workspace, we can keep the file .gitignored and force-add it to the working tree.

bterlson commented:
I have a vague worry that there are other things in vscode doing npm install for you, but talking with @ramya-rao-a this may be the only place. I like disabling it in settings.json. I also kinda like the hack you mention to allow updating of local settings! It lets me configure my extensions I love that nobody else has 😊 But, I'll leave that up to you, I'm ok dealing with untracked changes as well.

mikeharder commented on sdk/eventhub/testhub/commands/sendReceive.ts:
```diff
@@ -3,7 +3,7 @@
 import { EventHubClient, EventPosition, EventData, delay } from "@azure/event-hubs";
 import { log, setCurrentCommand, randomNumberFromInterval } from "../utils/util";
-import * as uuid from "uuid/v4";
+import uuid from "uuid/v4";
```
Why was this change needed?

bterlson replied:
esModuleInterop getting enabled, most likely.

bsiegel replied:
It wasn't building for me at all with rollup without this change. I think the older version of rollup was okay with it, but the newer version uses esModules everywhere internally

mikeharder commented on sdk/eventhub/testhub/package.json:
```diff
@@ -2,27 +2,41 @@
+    "test": "true"
   }
-}
+}
```
NIT: Missing newline

mikeharder commented on sdk/keyvault/keyvault/package.json:
```diff
@@ -47,9 +49,15 @@
   },
   "sideEffects": false
-}
+}
```
NIT: Missing newline

mikeharder commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -75,18 +76,22 @@
+    "unit-node": "npm run build-test && mocha -t 120000 test-dist/index.js"
   }
-}
+}
```
NIT: Missing newline

bterlson commented:
FWIW, my experience is full builds are much faster as it seems to skip things that are already built.

mikeharder commented:
> FWIW, my experience is full builds are much faster as it seems to skip things that are already built.

I will run a few before/after perf tests.

bsiegel commented on common/scripts/rush-welcome.js:
```diff
@@ -0,0 +1,14 @@
+const y = `\u001B[33m`;
```
Nope, I wrote this one myself. It adds a helpful message if someone runs `npm install` within the repo. Rush default scripts as well as custom scripts are all meant to live in this directory.

bsiegel commented on common/scripts/sync-versions.js:
```diff
@@ -0,0 +1,137 @@
+var fs = require("fs");
```
Also no, this was actually created by @fearthecowboy - they're using Rush for powershell. I just swiped it (and made a few tweaks).

bsiegel commented on gulpfile.ts:
```diff
@@ -62,7 +62,7 @@ if (!fs.existsSync(dropFolderPath)) {
 }
 
-gulp.task('default', () => {
+gulp.task('default', async () => {
```
Gulp tasks must all return a Promise. In older versions of Gulp this generated a warning. In the one we are using now, it's an error. It appears this was just an oversight - all the other tasks are marked async.

bsiegel commented on sdk/eventhub/event-hubs/package.json:
```diff
@@ -78,20 +78,22 @@
-    "build-samples": "cd examples && tsc -p .",
+    "lint": "tslint -p . -c tslint.json",
+    "pack": "npm pack 2>&1",
+    "prebuild": "npm run clean",
```
Good question. `prebuild` is not documented here: https://docs.npmjs.com/misc/scripts so I'm not sure if it will be run automatically at all. It might just be a confusing name. I can track down the originator of this line and ask them.

bsiegel replied:
@mikeharder I just tested locally and this doesn't get invoked when "build" is run. So I guess it's just a confusing name. It seems to have been added to this particular library by @amarzavery in b153ade, but I see that same script in numerous packages. I'm inclined to just leave it since it doesn't seem to affect the build (at least with Rush).

bsiegel commented on sdk/eventhub/event-hubs/package.json:
```diff
@@ -78,20 +78,22 @@
   ],
   "scripts": {
-    "tslint": "tslint -p . -c tslint.json",
+    "build-browser": "tsc -p . && cross-env ONLY_BROWSER=true rollup -c 2>&1",
```
Rush treats any output to STDERR as though it's a warning, and it makes the builds show up as completing with lots of warnings when some tools that we use just use STDERR to emit status messages. It kind of sucks, but that was the best way to silence those "warnings" cross-platform. This is documented in the README, under "Onboarding a new library".

bsiegel commented on sdk/eventhub/event-processor-host/package.json:
```diff
@@ -69,20 +72,22 @@
+    "unit-node": "npm run build-test && mocha -t 120000 test-dist/index.js"
   }
-}
+}
```
Yeah I need to figure out - some tools that process the package.json insist on adding them, and others insist on removing them. I don't care which we follow, but we should be consistent.

mikeharder replied:
My vote is to end with a trailing newline, since I don't like the red "no newline" symbol in github :).

bsiegel commented:
> 1. Should the Azure Pipelines YAML be updated to use the Rush commands for building, testing, etc?

Yep, I am planning to do that in a followup PR.

weshaggard commented on common/config/rush/command-line.json:
```diff
@@ -0,0 +1,278 @@
+/**
```
Sorry I accidentally posted this to another PR, not sure what happened, and I know this is already merged but I'm still curious about my question below.

Is there any hard rules about these files going into a folder called common? I ask mostly as for some of the other languages we have been putting similar engineering system type files under an "eng" directory.
